src/sinkhorn.rs

```diff
@@ -0,0 +1,21 @@
+use ndarray::{Array1, Array2, Axis};
+use float_ord::FloatOrd;
+
+pub fn sinkhorn(a: Array1< f32 >, b: Array1< f32 >, cost: Array2< f32 >, reg: f32) -> Array2< f32 >
+{
+    let n = cost.nrows();
+    let m = cost.ncols();
+    let epsilon = FloatOrd(1e-8);
+
+    let mut P = cost.mapv( |x| (-x / reg).exp());
+    P /= P.sum();
+
+    let mut u = Array1::zeros(n);
+    while (u.clone()-P.sum_axis(Axis(1))).mapv( |x: f32| FloatOrd(x.abs())).iter().max().unwrap() > &epsilon
+    {
+        u = P.sum_axis(Axis(1));
+        P = P.clone() * (a.clone() / u.clone()).broadcast((n, m)).unwrap();
```

Comment:
These clones seem really inefficient, I wonder if there's an easier way to do this

Reply:
do you need all of these clones?

I probably don't fully understand what's going on but my current impression is that these are cloned to retain access to values through loop iterations (i.e. the values are not 'moved') so that later iterations can still access those values.

I tried changing P.clone() to P on this line and it still seems to compile and run ok...
- it's ok if the old P value is 'moved' as we are moving it into a new P value assignment

Also seemed ok to change u.clone() to u inside the while statement
- in this case there is no assignment so the old u value is not moved

Is the 'move' consideration the only reason why clones are needed here?